packages/apputils/src/widgettracker.ts

```diff
@@ -0,0 +1,352 @@
+ *
+ * @typeparam T - The type of widget being tracked. Defaults to `Widget`.
+ *
+ * #### Notes
+ * The API surface area of this concrete implementation is substantially larger
+ * than the widget tracker interface it implements. The interface is intended
+ * for export by JupyterLab plugins that create widgets and have clients who may
+ * wish to keep track of newly created widgets. This class, however, can be used
+ * internally by plugins to restore state as well.
+ */
+export class WidgetTracker<T extends Widget = Widget>
+  implements IWidgetTracker<T>, IRestorable<T> {
+  /**
+   * Create a new widget tracker.
+   *
+   * @param options - The instantiation options for a widget tracker.
+   */
+  constructor(options: WidgetTracker.IOptions) {
+    const focus = (this._focusTracker = new FocusTracker());
+    const pool = (this._pool = new RestorablePool(options));
+
+    this.namespace = options.namespace;
+
+    focus.currentChanged.connect((_, current) => {
+      if (current.newValue !== this.currentWidget) {
+        pool.current = current.newValue;
+      }
+    }, this);
+
+    pool.added.connect((_, widget) => {
+      this._widgetAdded.emit(widget);
+    }, this);
+
+    pool.currentChanged.connect((_, widget) => {
+      if (widget === null && focus.currentWidget) {
+        pool.current = focus.currentWidget;
+        return;
+      }
+      this.onCurrentChanged(widget);
+      this._currentChanged.emit(widget);
+    }, this);
+
+    pool.updated.connect((_, widget) => {
+      this._widgetUpdated.emit(widget);
+    }, this);
+  }
+
+  /**
+   * A namespace for all tracked widgets, (e.g., `notebook`).
+   */
+  readonly namespace: string;
+
+  /**
+   * A signal emitted when the current widget changes.
+   */
+  get currentChanged(): ISignal<this, T | null> {
+    return this._currentChanged;
+  }
+
+  /**
+   * The current widget is the most recently focused or added widget.
+   *
+   * #### Notes
+   * It is the most recently focused widget, or the most recently added
+   * widget if no widget has taken focus.
+   */
+  get currentWidget(): T | null {
+    return this._pool.current || null;
+  }
+
+  /**
+   * A promise resolved when the tracker has been restored.
+   */
+  get restored(): Promise<void> {
+    return this._pool.restored;
+  }
+
+  /**
+   * The number of widgets held by the tracker.
+   */
+  get size(): number {
+    return this._pool.size;
+  }
+
+  /**
+   * A signal emitted when a widget is added.
+   *
+   * #### Notes
+   * This signal will only fire when a widget is added to the tracker. It will
+   * not fire if a widget is injected into the tracker.
+   */
+  get widgetAdded(): ISignal<this, T> {
+    return this._widgetAdded;
+  }
+
+  /**
+   * A signal emitted when a widget is updated.
+   */
+  get widgetUpdated(): ISignal<this, T> {
+    return this._widgetUpdated;
+  }
+
+  /**
+   * Add a new widget to the tracker.
+   *
+   * @param widget - The widget being added.
```

Comment:
I'm not sure that note makes sense. It basically says "you can use your widget now" ... which you probably always assume when you have a handle to the object. Can you think of a situation where you need more insight into what is happening in this `async` function?

Reply:
This came up when I was working on the `no-floating-promises` stuff. Basically, the question is "when is it safe to use the various querying operations on the tracker?". As soon as it is set, you can start doing things like `has` for `find` or `filter`. We were doing that in the tests without `await`ing the promise, and it wasn't necessarily wrong to do so, since we didn't really care when the promise resolved to do the in-memory stuff.